lib/boavizta.ts

```diff
@@ -0,0 +1,189 @@
+        }
+    }
+
+    protected formatResponse(response: any): { [key: string]: any } {
+        let m = 0;
+        let e = 0;
+        if ('impacts' in response.data) {
+            m = response.data['impacts']['gwp']['manufacture'] * 1000
+            e = response.data['impacts']['pe']['use'] / 3.6;
+        } else if ('gwp' in response.data && 'pe' in response.data) {
+            m = response.data['gwp']['manufacture'] * 1000
+            e = response.data['pe']['use'] / 3.6;
+        }
+        return {m, e};
+    }
+
+    transformToBoaviztaUsage(duration: any, metric: any) {
+        let usageInput: { [key: string]: any } = {
+            "hours_use_time": this.convertToHours(duration),
+            "time_workload": metric * 100.0,
+        }
+        usageInput = this.addLocationToUsage(usageInput);
+        return usageInput;
+    }
+
+    async calculate(observations: object | object[] | undefined = undefined): Promise<object> {
+        let mTotal = 0;
+        let eTotal = 0;
+        if (Array.isArray(observations)) {
+            if (observations.length === 0) {
+                throw new Error("Parameter Not Given: Missing observations parameter")
+            }
+            for (const usageRaw of observations) {
+                const usageResult = await this.calculateUsageForItem(usageRaw);
+                mTotal = usageResult.m;
+                eTotal += usageResult.e;
+            }
+        } else {
+            let m = 0, e = 0;
+            if (observations !== undefined) {
+                const usageResult = await this.calculateUsageForItem(observations);
+                m = usageResult.m;
+                e = usageResult.e;
+            } else {
+                throw new Error("Parameter Not Given: Missing observations parameter")
+            }
+            mTotal += m;
+            eTotal += e;
+        }
+        return {
+            "e": eTotal,
+            "m": mTotal
+        };
+    }
+
+    protected async calculateUsageForItem(usageRaw: { [key: string]: any }) {
+        if ('datetime' in usageRaw && 'duration' in usageRaw && this.metricType in usageRaw) {
+            const usageInput = this.transformToBoaviztaUsage(usageRaw['duration'], usageRaw[this.metricType]);
+            return await this.fetchData(usageInput) as IBoaviztaUsageSCI
+        } else if ('hours_use_time' in usageRaw && 'time_workload' in usageRaw) {
+            const usageRawWithLocation = this.addLocationToUsage(usageRaw);
+            return await this.fetchData(usageRawWithLocation) as IBoaviztaUsageSCI
+        } else {
+            throw new Error("Invalid Input: Invalid observations parameter")
+        }
+    }
+
+    addLocationToUsage(usageRaw: { [key: string]: any }) {
+        if (this.sharedParams !== undefined && 'location' in this.sharedParams) {
+            usageRaw['usage_location'] = this.sharedParams['location']
+        }
+        return usageRaw;
+    }
+}
+
+export interface IBoaviztaUsageSCI {
```

Comment:
Is this used anywhere?

Reply:
Yes. This is used as a function return type.